jOOL-java-8/src/main/java/org/jooq/lambda/tuple/Tuple2.java

```diff
@@ -264,15 +276,76 @@ public final Tuple0 skip2() {
      * </pre></code>
      */
     public static final <T extends Comparable<? super T>> Optional<Tuple2<T, T>> intersect(Tuple2<T, T> left, Tuple2<T, T> right) {
-        if (overlaps(left, right))
+        if (overlaps(left, right)) {
+            if (!left.finite() || !right.finite()) {
+                // todo?
+                throw new UnsupportedOperationException("to be done in another commit!");
+            }
             return Optional.of(new Tuple2<>(
-                left.v1.compareTo(right.v1) >= 0 ? left.v1 : right.v1,
-                left.v2.compareTo(right.v2) <= 0 ? left.v2 : right.v2
+                    left.v1.compareTo(right.v1) >= 0 ? left.v1 : right.v1,
+                    left.v2.compareTo(right.v2) <= 0 ? left.v2 : right.v2
             ));
-        else
+        } else
             return Optional.empty();
     }
 
+    /**
+     * Whether the given values are included in the tuple. Assumes inclusiveness.
+     * <p>
+     * <code><pre>
+     * // true
+     * contains(range(1, 3), 1, 3)
+     * contains(range(2, null), 5)
+     *
+     * // false
+     * contains(range(1, 3), 2, 4)
+     * contains(range(2, null), 1)
+     * </pre></code>
+     */
+    @SafeVarargs
+    public static final <T extends Comparable<? super T>> boolean contains(Tuple2<T, T> tuple2, T... vals) {
+        return Arrays.stream(vals).map(v -> {
+            if (tuple2.v1 != null && tuple2.v1.compareTo(v) > 0)
+                return false; // value is below the lower bound
+            if (tuple2.v2 != null && tuple2.v2.compareTo(v) < 0)
+                return false; // value is above the upper bound
+            return true;
+        }).reduce(Boolean::logicalAnd).orElse(false);
```

Comment:
I'm not sure we need a multi-val `contains()` method, but in case we do, it would be much better to "fail" early, instead of reducing the entire stream. In imperative logic, this would be much simpler to code (and faster, given we'll need less stack frames and have more inlining potential). But if you want to use a stream, I think you could use:

```java
}).filter(v -> !v).findAny().orElse(true);
```

Reply:
That's a good advice thank you.
Stream wasn't a requirement here, it was just used in the other method `Range#contains(Range other)` so I could call this `contains` method with both bound instead of calling it twice.

Reply:
Indeed, nice observation. Over the past years, I've (drastically) changed my mind about the utility of `Stream` usage, especially in library internals. It's still OK to use it in business logic, where it may indeed improve readability over imperative logic (sometimes it doesn't improve it), at a negligible performance overhead, because business logic is much more I/O bound (e.g. databases) than CPU bound. But in libraries, the stream overhead can be a killer.